### PR TITLE
Integration update

### PR DESCRIPTION
**Pull request by nfurfaro**

### CODE REVIEW

File: apps/utils/src/deploy.rs
Comment (supiket) [inline apps/utils/src/deploy.rs:8]: I suggest a function that registers both a contract and its programs for app utils. In the node, we are taking a different approach of registering contracts and programs separately, but I think for apps it will be most useful to have a function like:

```suggestion    
pub async fn register_contract_and_programs(
    builder_conn: &ConnectionPool,
    contract_registry: &PredicateAddress,
    program_registry: &PredicateAddress,
    contract: &Contract,
    programs: Vec<Program>,
) -> anyhow::Result<ContentAddress> {
    let solutions = register_contract_solution(contract_registry.clone(), contract)
        .into_iter()
        .chain(
            programs
                .iter()
                .map(|p| register_program_solution(program_registry.clone(), p)),
        )
        .collect();
    let solution_set = Arc::new(SolutionSet { solutions });
    let r = builder_conn
        .insert_solution_set_submission(
            solution_set,
            SystemTime::now().duration_since(UNIX_EPOCH).unwrap(),
        )
        .await?;
    Ok(r)
}
```

Even if that is not the way to go, at the very least I suggest not constructing the default big bang multiple times. This can be achieved by passing necessary big bang fields to relevant functions.

File: apps/utils/src/deploy.rs
Comment (supiket) [inline apps/utils/src/deploy.rs:1]: I think that we can replace the "deploy" term with "register" overall. With the note that there are a few mentions of "deployment" left in the node, mostly in comments though. Thoughts?

**Review by supiket — CHANGES_REQUESTED** [review]
Huge work on this PR, congrats! 🚀 

Submitting my initial review so that I don't keep you waiting until I review the rest.

File: apps/utils/src/deploy.rs
Comment (nfurfaro) [reply apps/utils/src/deploy.rs:1]: yeah I'm not sure.
Even though I guess "register" is more accurate, when building an app or writing tests, the mental model of "deployment" is probably more familiar?
Maybe not enough reason to go that route, but worth considering.

File: apps/utils/src/deploy.rs
Comment (nfurfaro) [reply apps/utils/src/deploy.rs:8]: I feel like devs shouldn't need to even think about deploying programs... it's almost an implementation detail (by this I mean that in Pint, there's no `program` keyword; we just write contracts, with predicates, constraints and state reads).
I do think one function should achieve both deployment of the contract and its associated programs, just not sure if devs should even have to know about the program part.

Thoughts?

File: apps/utils/src/deploy.rs
Comment (supiket) [reply apps/utils/src/deploy.rs:1]: I see, but I don't see a reason why the single term "register" wouldn't be enough for this action. It is used in the big bang and many other places already, "deploy" feels a bit outdated in practice even if not in semantics

File: apps/utils/src/deploy.rs
Comment (supiket) [reply apps/utils/src/deploy.rs:8]: I think it is not unreasonable to expect devs to know that:
- Solution sets contain solutions that point to predicates via their addresses in predicate registry contract's state. Those addresses should be __valid__ predicate addresses.
- Similarly, predicates point to programs via their addresses in program registry contract's state. Those addresses should be __valid__ program addresses.

Note: __Valid__ here means that the corresponding predicate / program is either already registered, or is being registered in the solution that attempts to use it.

But I see your point that pint abstracts programs away. The protocol is built on them though 🤔

File: crates/pint-submit/src/main.rs
Comment (mitchmindtree) [inline crates/pint-submit/src/main.rs:42]: Just noticing that this behaviour is a bit different to the previous behaviour. Now we only allow receiving an individual `Solution`, whereas previously we accepted a set of solutions. I can see it being useful to support both! For now, I'll change this to `--solution-set` and `SolutionSet`, to match the original behaviour more closely, but I'll open an issue to add support for submitting individual solutions (which might also help to simplify the "Deploy and Solve" chapter a little).

**Review by mitchmindtree — COMMENTED** [review]
*copied from slack*

Just noticing the `docs/code/front-end.sh` script causes a bit of a circular dependency within the repo itself.

This is because it uses `cargo add` to install the necessary versions of each tool including those from essential-integration. The problem with this is that the essential-integration tools cannot be published until the PR lands, but the PR can never land while it is blocked by this `./front-end.sh` test.

I think we'll want to do something like the following:

1. Add a final commit that updates the counter examples and any remaining mentions of decision variables etc.
2. Land this PR with the `./front-end.sh` test disabled temporarily so that we can publish the crates.
3. Work out some other solution for testing the `./front-end.sh`, or alternatively come up with some better way of specifying these dependencies than using `cargo add` (e.g. using something like this https://github.com/essential-contributions/essential-integration/issues/124).

I'll put together a commit for 1. now!

Edit: Opened https://github.com/essential-contributions/essential-integration/issues/145.

File: Cargo.lock
Comment (mitchmindtree) [inline Cargo.lock:918]: Just noticed these old deps are still here. It looks like they're pulled in via `essential-debugger`, so we might need to give this an update too, I'll have a look into this now.

Comment (mitchmindtree) [reply Cargo.lock:918]: Instead, I've removed the dep from `essential-app-utils` as it was unused.

We already bring the debugger CLI in via nix input and provide it to the developer shell. We should update it, but it looks like it hasn't been updated in a few months so might deserve its own PR.

File: crates/pint-submit/src/main.rs
Comment (mitchmindtree) [reply crates/pint-submit/src/main.rs]: Addressed in #144

**Review by mitchmindtree — APPROVED** [review]
Should be ready to go!

File: .github/workflows/ci.yml
Comment (mitchmindtree) [inline .github/workflows/ci.yml:68]: Issue opened at https://github.com/essential-contributions/essential-integration/issues/145

**Review by nfurfaro — COMMENTED** [review]
Nice work @mitchmindtree ! 
Just needs @supiket 's approval.

**Review by supiket — APPROVED** [review]

**Pull request merged**
